Shapes: v2: add SHACL shapes for Core

DylanVanAssche opened the pull request:
Core specification has now a proper ontology and diagram, add corresponding SHACL shapes:

- RML Triples Map
- RML Subject Map
- RML Predicate Map
- RML Object Map
- RML Predicate Object Map
- RML Language Map
- RML Graph Map
- RML Datatype Map

The complete shape is available in `core.ttl` which covers the whole RML-Core specification.

`generate-shape.py` takes all the shapes and generates a single one for RML Core. This is useful for SHACL validators where you can just load a single Shape file.

Note that these shapes do not have any test cases yet, these can be added later.

pmaria commented on Shapes/shapes/v2/language_map.ttl:
```diff
@@ -0,0 +1,140 @@
+    Literal from a logical table row or iterator.
+    """ ;
+    sh:message """
+    rr:language must be a valid language tag according to BCP47 and may
```
```suggestion
    rml:language must be a valid language tag according to BCP47 and may
```

DylanVanAssche replied:
Will fix, thanks!

pmaria reviewed (commented):
Not an in-depth review, but some comments:
* Since the term map shapes are not closed anymore, we can now simplify the term map shapes significantly by moving some property shapes to the level of an `ExpressionMap` shape.
Also, `sh:ignoredProperties`  could be removed
* Some shapes are still closed. Is there a need for that?
* As for extensibility of the core shape, already an interesting one is how to add a new expression type in a property shape with a construct like `sh:path [sh:alternativePath (rml:reference rml:template rml:constant)] ;`

DylanVanAssche commented:
> Since the term map shapes are not closed anymore, we can now simplify the term map shapes significantly by moving some property shapes to the level of an ExpressionMap shape.

I haven't done that for the reason that we cannot re-use them across specifications without depending on the core. Each set of shapes is stand-alone. Refactoring them could be done, but won't benefit us much until you can re-use them across specifications.

> Some shapes are still closed. Is there a need for that?

No, that's just forgotten.

> As for extensibility of the core shape, already an interesting one is how to add a new expression type in a property shape with a construct like sh:path [sh:alternativePath (rml:reference rml:template rml:constant)] ;

Well you need to add it there + the shape validating the new expression type.

andimou commented on Shapes/shapes/v2/datatype_map.ttl:
```diff
@@ -0,0 +1,139 @@
+
+schema:RMLDatatypeMapShape
+    a sh:NodeShape ;
+    sh:ignoredProperties (rdf:type) ;
```
why is this property ignored?

DylanVanAssche replied:
Should be dropped, it is from `sh:closed` as @pmaria pointed out, will fix :)

andimou commented on Shapes/shapes/v2/datatype_map.ttl:
```diff
@@ -0,0 +1,139 @@
+
+    # Exactly one rml:template or one rml:reference or one rml:constant is
+    # required
+    sh:property [
```
I think it would be easier if we give this property shape a URI.

I think that the DatatypeMap is an ExpressionMap so in principle its constraints should follow the constraints of the Expression Map.

DylanVanAssche replied:
There's no inheritance on purpose to keep things readable and understandable which is also why I never use IRIs.

andimou commented on Shapes/shapes/v2/graph_map.ttl:
```diff
@@ -0,0 +1,138 @@
+    ] ;
+
+    # rml:reference
+    sh:property [
```
that's the `ReferencePropertyShape` mentioned in my other comment right?

```schema:RMLGraphMapShape sh:property sh:and ( :ReferencePropertyShape :LiteralPropertyShape).

:ReferencePropertyShape sh:path rml:reference ; sh:node ReferenceNodeShape.

:LiteralPropertyShape sh:path rml:termType; sh:in (rml:IRI rml:BlankNode)`

(I hope it doesn't contain mistakes but it might need some refinement)

andimou commented on Shapes/shapes/v2/language_map.ttl:
```diff
@@ -0,0 +1,140 @@
+    """ ;
+
+    # Exactly one rml:template or one rml:constant is required
+    sh:property [
```
that repeats again, I just mention to further support my comment that it might be better to give IRIs to the shapes and reuse them instead of redefining them again and again.

andimou commented on Shapes/shapes/v2/ref_object_map.ttl:
```diff
@@ -0,0 +1,85 @@
+        sh:nodeKind sh:BlankNodeOrIRI ;
+
+        # rml:parent
+        sh:property [
```
this is good for the shortcut but we might need to introduce a shape for the ParentMap and ChildMap as well which are the same as what rml:languageMap is to rml:language

andimou reviewed (commented):
@DylanVanAssche thank you for preparing all these!

A few high level comments:
* I think it's better if we have IRIs for the property shapes instead of using blank nodes. That would help us to reuse shapes as I currently have the feeling that we regenerate the same property shapes in multiple places.
* I have the impression that certain properties need to go to the Expression Map and then reuse the Expression Maps for the different TermMaps, DatatypeMap/LanguageMap/ParentMap/ChildMap.
* I think that we might need to refine some of the shapes as I think that currently all combinations of expression maps and term types are possible in all places but this is not desired. What we want to do is to allow certain combinations. For instance, currently the `rml:reference` property is independent of the `rml:termType` property. so we might end up with a literal for subject. However, what we want to allow are certain combinations: 

- * SM: whatever expression map with IRI or blank node term type
- * PM: whatever expression map with IRI term type
- * OM: whatever expression map with whatever term type
- * LanguageMap (LM), ParentMap, ChildMap: 
whatever expression map with Literal term type
- * DatatypeMap (DM): whatever expression map with IRI term type

* There might be a small mistake with the sh:language, I left a detailed comment.

Overall, some of my more detailed comments repeat as several shapes are redefined.

If something is not clear, let me know!

DylanVanAssche commented on Shapes/shapes/v2/graph_map.ttl:
```diff
@@ -0,0 +1,138 @@
+
+    # Exactly one rml:template or one rml:constant is required
+    sh:property [
+        sh:path [sh:alternativePath (rml:template rml:constant)] ;
```
Oh this seems to be an old thing, back when `rml:reference` was not supported for this. (Or it was, but the spec back then, pre-W3C, did not make things clear).

DylanVanAssche commented:
@andimou I addressed your comments, but followed the way as defined in the SHACL specification by using `sh:and`.
These changes make it way shorter indeed.

So in each shape I override the list of `rml:IRI`, `rml:BlankNode`, `rml:Literal` depending on the Term Map. This way, you can only use `rml:IRI` as term type in a Subject Map or Predicate Map, while all of them in an Object Map.

Let me know if you have more comments.

DylanVanAssche commented:
Imported the RML Test Cases from https://github.com/kg-construct/rml-test-cases
They are not fully perfect yet, but are sufficient for validation with SHACL.
